docs/development/design/build-images.rst

```diff
@@ -48,131 +45,242 @@ This produced issues to people pinning their images to any of these names becaus
 
-Specifying extra user's dependencies
-------------------------------------
 
-Different users may have different requirements. We were already requested to install
-``swig``, ``imagemagick``, ``libmysqlclient-dev``, ``lmod``, ``rust``, ``poppler-utils``, etc.
+Specifying extra languages requirements
+---------------------------------------
 
-People with specific dependencies will be able to install them as APT packages or as extras
-using ``.readthedocs.yaml`` config file. Example:
+Different users may have different requirements.
+People with specific language dependencies will be able to install them by using ``.readthedocs.yaml`` config file.
+Example:
 
 .. code:: yaml
 
+   version: 3
    build:
-     image: ubuntu20
-     python: 3.9
-     system_packages:
-       - swig
-       - imagemagick
-     extras:
-       - node==14
-       - rust==1.46
+     os: ubuntu20
+     languages:
+       python: "3.9"  # supports "pypy3", "miniconda3" and "mambaforge"
+       nodejs: "14"
+       rust: "1.54.0"
+       golang: "1.17"
 
 Important highlights:
 
-* users won't be able to use custom Ubuntu PPAs to install packages
-* all APT packages installed will be from official Ubuntu repositories
-* not specifying ``build.image`` will pick the latest OS image available
-* not specifying ``build.python`` will pick the latest Python version available
-* Ubuntu 18 will still be available via ``stable`` and ``latest`` images
-* all ``node`` (major) pre-compiled versions on ``nodenv`` are available to select
-* all ``rust`` (minor) pre-compiled versions on ``rustup`` are available to select
-* knowing exactly what packages users are installing,
-  could allow us to prebuild extra images: ``ubuntu20-py37+node14``
+* do not treat Python language different from the others (will help us to support other non-Python doctools in the future)
+* specifying ``build.languages.python: "3"`` will use Python version ``3.x.y``, and may differ between builds
+* specifying ``build.languages.python: "3.9"`` will use Python version ``3.9.y``, and may differ between builds
+* specifying ``build.languages.nodejs: "14"`` will use nodejs version ``14.x.y``, and may differ between builds
+* if no full version is declared, the latest available on ``asdf`` will be installed
+* specifying minor language versios is not allowed (e.g. ``3.7.11``)
+* not specifying ``build.os`` will make the config file parser to fail
+* not specifying ``build.languages`` will make the config file parsing to fail (at least one is required)
+* specifying only ``build.languages.nodejs`` and using Sphinx to build the docs, will make the build to fail
+* ``build.image`` is incompatible with ``build.os`` or ``build.languages`` and will produce an error
+* Ubuntu 18 will still be available via ``stable`` and ``latest`` images, but not in new ones
+* a subset (not defined yet) of ``python``, ``nodejs``, ``rust`` and ``go`` versions on ``asdf`` are available to select
```

Comment:
What does this mean? We are going to specify somehow the versions that users can install? Is this managed by us or some external tool?

Reply:
This is not decided yet. I added it here to discuss with the team.

We could support:

- install any version available on `asdf list all <language>` or,
- maintain a list/regex of allowed version from the previous list

IMO, there is no need to expose _all the versions available to install_ and we should expose a fixed list that we have control and we are sure they provide binaries (to avoid compiling the other languages nor creating cache for the others).